### PR TITLE
Add a more sophisticated validation method for C++ test.

### PR DESCRIPTION
**Pull request by csarofeen**
Validation tolerances vary based on how many values have been reduced. I empirically gathered tolerances as a function of number of elements reduced using sum(add(t0, t1)). I then apply this when we have fusion with reductions in them to estimate what the validation tolerance should be. The experiments were run with [-1, 1] values, so this should be a good method when doing reductions as long as the values reduced are [-2, 2].

### CODE REVIEW

File: test/cpp/jit/test_gpu_validator.h [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+
+ private:
+  ReductionSizeMapper(Fusion* fusion, ExpressionEvaluator& expr_eval)
+      : expr_eval_(expr_eval) {
+    // Initialize input values
+    for (auto inp : fusion->inputs()) {
+      if (inp->isA<TensorView>()) {
+        auto tv = inp->as<TensorView>();
+        // Shouldn't have any reductions, but run it through analysis anyways.
+        reduction_map[tv] = getReductionSize(tv);
```
Comment (shmsong) [inline test/cpp/jit/test_gpu_validator.h:147]: Quick question: we currently don't seem to have input tensors with reduction axes. Do we have more cases where we need to call `getReductionSize` on fusion inputs other than when we start partitioning fusions?

File: test/cpp/jit/test_gpu_validator.h [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+          // Grab the next entry up so we have some margin
+          abs_tol = sum_tolerance_entry[entry + 1][1];
+        } else {
+          // If we hit the end of the list, return twice the max error we
+          // measured
+          abs_tol = sum_tolerance_entry[sum_tolerance_entry.size() - 1][1] * 2.;
+        }
+        // Relative tol we're going to set to 1% of abs tol just for
+        // a small margin of rel error.
+        return {abs_tol, abs_tol * 0.01};
```
Comment (shmsong) [inline test/cpp/jit/test_gpu_validator.h:115]: from PyTorch: [TORCH.ALLCLOSE](https://pytorch.org/docs/stable/generated/torch.allclose.html)
```
∣input−other∣ ≤ atol+rtol×∣other∣
```
I don't believe `rtol` is meant to scale with `atol` .

Also I use to think having more reduction elements shouldn't require relaxing `rtol`.

But could be I'm not understanding this scheme.

File: test/cpp/jit/test_gpu_validator.h (hunk 1 above)
Comment (csarofeen) [reply test/cpp/jit/test_gpu_validator.h:147]: I was thinking so and wanted it to be compatible with that yes.

File: test/cpp/jit/test_gpu_validator.h (hunk 2 above)
Comment (csarofeen) [reply test/cpp/jit/test_gpu_validator.h:115]: I was going back and forth on this, was thinking to have rtol as 0 and just use atol, but that seemed strange to me, so just put a small amount of rtol.

**Review by shmsong — APPROVED** [review]
Looks great. This is very helpful in validating reduction heavy fusion. Thanks!

**Pull request merged**